### PR TITLE
Correct live/ready checks to consider special host values

### PR DESCRIPTION
**Pull request by steve-chavez**
Fixes https://github.com/PostgREST/postgrest/issues/2174.

Took a different route from the one suggested in https://github.com/PostgREST/postgrest/issues/2174#issuecomment-1045978929. I avoided including the [streaming-commons](https://hackage.haskell.org/package/streaming-commons) dependency and instead just handled the special host values in the admin server logic by stealing some of the code from [bindPortGenEx](https://hackage.haskell.org/package/streaming-commons-0.2.2.4/docs/src/Data.Streaming.Network.html#bindPortGenEx). That seemed like a simpler change.

### TODO

- [x] Refactor
- [x] Make tests green

### CODE REVIEW

**Comment by steve-chavez** [comment]
localhost is somehow not resolving.. not sure why.

**Comment by wolfgangwalther** [comment]
Hm. I'm not 100% convinced by this. When I run multiple instances of PostgREST on the same port, but different interfaces, one instance might go down, but the healthcheck could still show 200 OK, because it connects to the other instance instead.

So this has two problems:
- It tries multiple addresses, which might lead to the above.
- It determines those addresses at request-time, i.e. when the request to the live endpoint is made. At this time there might be other interfaces with other IPs.

A more reliable approach would be to run this once, when the main app port is assigned. And then only take the first available interface. But that will quickly lead to the same solution as mentioned in https://github.com/PostgREST/postgrest/issues/2174#issuecomment-1045978929.

File: test/io/test_io.py [hunk 1]
```diff
@@ -883,6 +884,17 @@ def test_admin_live_dependent_on_main_app(defaultenv):
         response = postgrest.admin.get("/live")
         assert response.status_code == 503
 
+def test_admin_works_with_host_special_values(defaultenv):
+    "Should get a success from the admin live and ready endpoints when using special host values for the main app"
+
+    for host in ["*4", "!4", "*6", "!6", "*"]:
```
Comment (wolfgangwalther) [inline test/io/test_io.py:890]: This is not how multiple almost identical test-cases are created in pytest. Have a look at some of the other test cases in this file that use `@pytest.mark.parametrize` to achieve it in a better way.

File: test/io/test_io.py [hunk 2]
```diff
@@ -140,16 +140,17 @@ def dumpconfig(configpath=None, env=None, stdin=None):
...
     env = env or {}
     env["PGRST_DB_POOL"] = "1"
     env["PGRST_DB_POOL_TIMEOUT"] = "1"
+    host = host or "localhost"
 
     with tempfile.TemporaryDirectory() as tmpdir:
         if port:
             env["PGRST_SERVER_PORT"] = str(port)
-            env["PGRST_SERVER_HOST"] = "localhost"
+            env["PGRST_SERVER_HOST"] = host
```
Comment (wolfgangwalther) [inline test/io/test_io.py:153]: Right you change the default host for all tests that don't specify `host`.

To not accidentally change unrelated things, it'd be better to do:

```suggestion
            if host:
                env["PGRST_SERVER_HOST"] = host
```

And then you don't need the `host = host or ...` line above.

**Comment by wolfgangwalther** [comment]
@laurenceisla There's a lot of unrelated changes that are already on `main` in here. I think you need to rebase properly to get rid of them; I think you somehow merged those in without rebase.

**Comment by laurenceisla** [comment]
> @laurenceisla There's a lot of unrelated changes that are already on `main` in here. I think you need to rebase properly to get rid of them; I think you somehow merged those in without rebase.

Fixed, I reverted it to the state before I commited. I will push the tests again after I fix them with the revisions you gave.

**Comment by steve-chavez** [comment]
> When I run multiple instances of PostgREST on the same port, but different interfaces, one instance might go down, but the healthcheck could still show 200 OK, because it connects to the other instance instead

Hm, this seems like an edge case but yeah it seems it can happen. In the case you have a machine with multiple network interfaces you'd be better using an specific host name for each postgrest service rather than the special values(!4, *, etc) though right? And in that case the edge case won't happen?

**Comment by wolfgangwalther** [comment]
> Hm, this seems like an edge case but yeah it seems it can happen. In the case you have a machine with multiple network interfaces you'd be better using an specific host name for each postgrest service rather than the special values(!4, *, etc) though right? And in that case the edge case won't happen?

True. Should be enough to point that out in the docs. Using multiple interfaces is only safe when setting specific values. And I think that's true for the main app as well, not just for the admin app.

---

This is one of the MRs where we would really benefit from the ability to build docker images for MRs, including upload to docker hub. I would really like to test this in my k8s setup, but don't know how.

**Comment by steve-chavez** [comment]
> True. Should be enough to point that out in the docs.

Cool. Then this is looking good to merge.

> This is one of the PRs where we would really benefit from the ability to build docker images for PRs, including upload to docker hub.

That would be nice, like the previews in readthedocs.. users could try new features too.. I guess the idea would be tagging an image for each CI run and the tag would be the PR's number..

**Pull request merged**